### PR TITLE
Handle xref monikers at build time

### PR DESCRIPTION
**Pull request by yufeih**
- Move moniker resolve logic from resolve time to build time, so the logic is executed only once for each uid resolve.
- Split internal xref logic to `XrefMapBuilder` and external xref logic to `XrefMapLoader`
- Fixes `ToExternalXrefSpec` error path by using `DeclaringFile`

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4783)

### CODE REVIEW

File: src/docfx/build/xref/IXrefSpec.cs [hunk 1]
```diff
@@ -11,7 +9,7 @@ internal interface IXrefSpec
 
         Document DeclairingFile { get; }
 
-        HashSet<string> Monikers { get; }
+        string[] Monikers { get; }
```
Comment (live1206) [inline src/docfx/build/xref/IXrefSpec.cs]: HashSet

File: src/docfx/build/xref/InternalXrefSpec.cs [hunk 2]
```diff
@@ -15,7 +15,7 @@ internal class InternalXrefSpec : IXrefSpec
 
         public Document DeclairingFile { get; set; }
 
-        public HashSet<string> Monikers { get; set; } = new HashSet<string>();
+        public string[] Monikers { get; set; } = Array.Empty<string>();
```
Comment (live1206) [inline src/docfx/build/xref/InternalXrefSpec.cs]: This should be `HashSet`, refer to usage at line 99 in `XrefMap.cs`

File: src/docfx/build/xref/XrefMap.cs [hunk 3]
```diff
@@ -47,170 +45,65 @@ public XrefMap(Context context, IReadOnlyDictionary<string, List<Lazy<IXrefSpec>
...
-            {
-                var (_, query, fragment) = UrlUtility.SplitUrl(spec.Href);
-                resolvedHref = UrlUtility.MergeUrl(spec.Href, query, fragment.Length == 0 ? "" : fragment.Substring(1));
-
-                name = spec.GetXrefPropertyValue("name");
-                displayPropertyValue = spec.GetXrefPropertyValue(displayPropertyName);
-            }
-            else
+            var spec = ResolveXrefSpec(uid, moniker);
+            if (spec == null)
```
Comment (live1206) [inline src/docfx/build/xref/XrefMap.cs]:
```suggestion
            if (spec is null)
```

File: src/docfx/build/xref/XrefMapBuilder.cs [hunk 4]
```diff
@@ -13,70 +12,30 @@ namespace Microsoft.Docs.Build
 {
     internal static class XrefMapBuilder
```
Comment (live1206) [inline src/docfx/build/xref/XrefMapBuilder.cs]: how about renaming this to `InternalXrefMapBuilder`?

File: src/docfx/build/xref/XrefMapLoader.cs [hunk 5]
```diff
@@ -13,17 +13,49 @@ internal class XrefMapLoader
     {
         private static byte[] s_uidBytes = Encoding.UTF8.GetBytes("uid");
 
-        public static List<(string, Lazy<IXrefSpec>)> Load(string filePath)
+        public static IReadOnlyDictionary<string, Lazy<ExternalXrefSpec>> Load(Docset docset)
```
Comment (live1206) [inline src/docfx/build/xref/XrefMapLoader.cs]: rename this class to `ExternalXrefMapLoader`?

File: src/docfx/build/xref/XrefMap.cs [hunk 6]
```diff
@@ -47,170 +45,65 @@ public XrefMap(Context context, IReadOnlyDictionary<string, List<Lazy<IXrefSpec>
...
-            }
-        }
 
-        private bool CheckOverlappingMonikers(IEnumerable<IXrefSpec> specsWithSameUid, out HashSet<string> overlappingMonikers)
-        {
-            bool isOverlapping = false;
-            overlappingMonikers = new HashSet<string>();
-            var monikerHashSet = new HashSet<string>();
-            foreach (var spec in specsWithSameUid)
+            if (!string.IsNullOrEmpty(moniker))
```
Comment (live1206) [inline src/docfx/build/xref/XrefMap.cs]: We are not covering all the scenarios in this PR. Check https://github.com/dotnet/docfx/pull/4621.
I am scheduling a meeting to have a requirement review firstly, then continue with this.

Comment (yufeih) [reply src/docfx/build/xref/XrefMap.cs]: yep, just keep existing code untouched

**Pull request merged**